proposals/0621-linear-constraints.rst

```diff
@@ -0,0 +1,1100 @@
+
+  bad :: C %1 => (Int, Int)
+  bad = (giveC useC, useC)
+
+  bad' :: C %1 => (Int, Int)
+  bad' = (giveC useC, 0)
+
+In ``bad``, if the leftmost ``useC`` uses the linear ``C`` from the
+function signature, then ``bad`` would be rejected, it must used the
+unrestricted ``C`` from ``giveC``. But in ``bad'`` it must use the
+linear ``C`` instead. So this would force the leftmost ``useC`` to
+make a guess. Instead we reject both ``bad`` and ``bad'``.
+
+Equality constraints
+~~~~~~~~~~~~~~~~~~~~
+
+Given equality constraints are used for unification *only
+if they are unrestricted* (correspondingly, equality constraints
+generated by the type inference algorithm are unrestricted, as they
+have always been).
+
+The reason for this is that there is no clear semantics to make use of
+a linear equality constraint ``a ~ b`` as part of the unification
+algorithm (it's not that reasoning about linear equality is
+meaningless, but a unification or congruence conversion for linear
+equality isn't obvious to come up with, if someone has we're not
+aware). Nor does it feel like a true limitation as there is no example
+where a linear equality would be useful. It's really not worth the
+bother of trying to find a solution.
+
+Dupable classes
+^^^^^^^^^^^^^^^
+
+A new module ``GHC.Constraint.Linear`` is introduced (inlined
+alternative: bikeshed names, including the module name).
+
+This module exposes the following:
+
+::
+
+  data LinearityToken :: ZeroBitType
+
+  consumeLinearityToken :: LinearityToken %1 -> (# #)
+  dup2LinearityToken :: LinearityToken %1 -> (# LinearityToken, LinearityToken #)
+
+  data DupableClassModifier = Dupable
+
+Class declaration can be annotated with the ``%Dupable``
+
+::
+
+   %Dupable class <ctx> => <head> where
+    <methods>
+
+Classes annotated with ``%Dupable`` must:
+
+- Have a single method
+- The method must be of type ``LinearityToken``
+
+Such a dupable type class can be used multiple times even if they are
+linear. *E.g.*::
+
+  class Foo where
+    foo' :: LinearityToken
+
+  foo :: Foo %1 => Int -> Int
+  foo = case foo' of { (# #) -> \x -> x }
+
+  dupes :: Foo %1 => (Int -> Int, Int -> Int)
+  dupes = (foo, foo)
+
+But, crucially, not passed to an unrestricted function::
+
+  rejected :: Foo %1 => Ur (Int -> Int)
+  rejected = Ur foo
+
+There are a lot of alternatives for the design of this feature, so see
+the *Alternatives* section for more thoughts.
+
+Examples
+--------
+
+Threading
+^^^^^^^^^
+
+We refer back now to the first example from the motivation section,
+which showed how writing a function which reads the first two elements
+of an array became a tedious exercise of threading our linear resource
+through the function. Using linear constraints, however, such a
+function can be written as:
+
+::
+
+   read2AndDiscard ::  (Read n, Write n) %1 => MArray a n -> (Ur a, Ur a)
+   read2AndDiscard arr =
+        let !(Box x)  = read arr 0
+            !(Box y)  = read arr 1
+            !()       = free arr
```

Comment:
Yes, I wonder if it's OK to use `do` notation instead, or whether that defeats the point of the proposal.

Reply:
It is possible to express so-called sequenced "let no rec" using qualified do notation with preserve of linearity: https://play.haskell.org/saved/5ahwqWY2

However, I don't think that this is a good idea to introduce do-notation based DSL in the proposal and then write all examples in terms of this DSL.

Reply:
@JakobBruenker 

> Here, I assume that order is determined by the order of bindings within the `let` block? Which makes sense, but is a bit surprising, since AFAIK until now, order of bindings within a `let` block has not been significant.

You can observe the order of strict bindings today

```haskell
let {!x = error "1"; !y = error "2"} in 0
```

GHC is rather careful around strict bindings. Though I can't think of a way to make it matter (that is, returning a different exception is considered to be the same semantics in Haskell; and I don't know whether we can concoct an example where the order of evaluation of let arguments will say, turn an exception into a non-exception).

The way the type-checker is specified in the paper also makes it so that there's a single solution (in Core, the ordering is simply the threading of a state token, so only the type-checker matters).

So you take away what you want from this: it's a possibly surprising twist on something that already exists.

Reply:
Interesting! Seems like it would be useful to spell out explicitly in the proposal that this is how the order is determined.

Reply:
(I think github will present this without context, so to be clear, this is a response in [an existing discussion thread](https://github.com/ghc-proposals/ghc-proposals/pull/621/files#r1412909224))

The example you provide here
```haskell
let {!x = error "1"; !y = error "2"} in 0
```
produces
```
*** Exception: 2
```
Doesn't that mean strict `let` bindings have the opposite evaluation order from what we would want?

(NB, the syntax I suggest in a different comment has the order one would expect:)

```haskell
ghci> if | !x <- error "1", !y <- error "2" -> ()
*** Exception: 1
```